### PR TITLE
.NET 5.0 with C# 9

### PR DESCRIPTION
**Pull request by NN---**
Merge after #124

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<Project>
+	<PropertyGroup>
+		<!-- Compile Include="x.cs" doesn't use LangVersion defined in the PropertyGroup -->
+		<LangVersion>9.0</LangVersion>
```
Comment (ig-sinicyn) [inline Directory.Build.props:4]: I'd prefer to keep the setting either in csproj or in \Build\Props\CodeJam.Default.props

File: Build/Props/CodeJam.Targeting.props [hunk 2]
```diff
@@ -11,8 +11,8 @@
 	<PropertyGroup>
 		<CopyMeMinimalTargetFrameworks>net461;netstandard2.0</CopyMeMinimalTargetFrameworks>
 		<CopyMeMinimalTestTargetFrameworks>net461;netcoreapp2.0</CopyMeMinimalTestTargetFrameworks>
-		<CopyMeTargetFrameworks>net472;net461;net45;net40;net35;net20;netstandard2.1;netstandard2.0;netstandard1.5;netstandard1.3;netcoreapp3.0;netcoreapp2.0;netcoreapp1.0</CopyMeTargetFrameworks>
-		<CopyMeTestTargetFrameworks>net48;net472;net471;net47;net461;net45;net40;net35;net20;netcoreapp3.0;netcoreapp2.1;netcoreapp1.1</CopyMeTestTargetFrameworks>
+		<CopyMeTargetFrameworks>net50;net472;net461;net45;net40;net35;net20;netstandard2.1;netstandard2.0;netstandard1.5;netstandard1.3;netcoreapp3.0;netcoreapp2.0;netcoreapp1.0</CopyMeTargetFrameworks>
```
Comment (ig-sinicyn) [inline Build/Props/CodeJam.Targeting.props:14]: We need to regenerate FW monikers in CodeJam.TargetingTests.EmitTargetingConstants

**Review by ig-sinicyn — CHANGES_REQUESTED** [review]

**Comment by ig-sinicyn** [comment]
@NN---  Thanks for the great work!)
I'm going to look more closely at the PR tomorrow.

File: Directory.Build.props (hunk 1 above)
Comment (NN---) [reply Directory.Build.props:4]: The problem is that it doesn't help.
Files such as `<Compile Include="JetBrains.Annotations.cs"/>` don't receive LangVersion 9.0.
I submitted a bug.

It it possible to include a file from Build\Props\... , but then there is no difference between setting it here or including.

Comment (NN---) [reply Directory.Build.props:4]: I simply didn't find other working solution.

Comment (ig-sinicyn) [reply Directory.Build.props:4]: Aha! Let's remove <LangVersion> from csproj files then

File: Build/Props/CodeJam.Targeting.props (hunk 2 above)
Comment (NN---) [reply Build/Props/CodeJam.Targeting.props:14]: Sure, will add there.

File: Directory.Build.props (hunk 1 above)
Comment (NN---) [reply Directory.Build.props:4]: Ok.
Btw, you can vote for the issue: https://developercommunity2.visualstudio.com/t/itemgroup-compile-include-doesnt-use-the-same-lang/1255582?from=email&preview=true

**Comment by ig-sinicyn** [comment]
As suggestion:
```cs
#pragma warning disable SYSLIB0012 // 'Assembly.CodeBase' is obsolete: 'Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location instead.

			var uri = new Uri(assembly.CodeBase);

#pragma warning restore SYSLIB0012
```
I think we should follow the recommendation and use CodeBase only when targeting to the net FW

**Comment by ig-sinicyn** [comment]
```
#if LESSTHAN_NET50
		[ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
#endif
```
I'm not sure but this looks kinda strange for me.
The line was ok for all prev targets, what changed with FW 5.0 ?

**Comment by ig-sinicyn** [comment]
As for me, everything else is ok.

One more time, a great thanks you! :)

**Comment by NN---** [comment]
@ig-sinicyn Next part is to 

> 
> 
> ```
> #if LESSTHAN_NET50
> 		[ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
> #endif
> ```
> 
> I'm not sure but this looks kinda strange for me.
> The line was ok for all prev targets, what changed with FW 5.0 ?

.NET 5.0 added ObsoleteAttribute to this :)
Before .NET 5.0 it didn't have any meaning in .NET Core, but was not obsolete.
https://github.com/dotnet/runtime/commit/329f0dbfa3a164ee2e3bf9f75b7641165995c799#diff-997b403ad145cdd790196073af16aa4df8b1fc688415ce38b1c8a788e438dd2a

**Comment by ig-sinicyn** [comment]
> .NET 5.0 added ObsoleteAttribute to this :)

Wow, nice catch!

File: Build/Props/CodeJam.Targeting.props (hunk 2 above)
Comment (NN---) [reply Build/Props/CodeJam.Targeting.props:14]: Actually didn't know it is autogenerated :)
Why not T4 as done in other places ?

**Comment by NN---** [comment]
@ig-sinicyn AppVeyor still don't have VS 16.8 , hence the preview version.
I think it is stable enough for CodeJam.
Once they update their image, we can switch back to release.

File: Build/Props/CodeJam.Targeting.props (hunk 2 above)
Comment (ig-sinicyn) [reply Build/Props/CodeJam.Targeting.props:14]: Let's fix it after this PR.

**Comment by NN---** [comment]
> 
> 
> As suggestion:
> 
> ```cs
> #pragma warning disable SYSLIB0012 // 'Assembly.CodeBase' is obsolete: 'Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location instead.
> 
> 			var uri = new Uri(assembly.CodeBase);
> 
> #pragma warning restore SYSLIB0012
> ```
> 
> I think we should follow the recommendation and use CodeBase only when targeting to the net FW

This is more complicated.
Actually the best way to do it in .NET Core with covering self-contained executable is : `Process.GetCurrentProcess().MainModule.FileName `

**Comment by NN---** [comment]
Fixed comments. Good enough to commit ?:)
I started adding nullability, but it requires C# 9.0 in order to work correctly with generic methods.

**Comment by ig-sinicyn** [comment]
> Fixed comments. Good enough to commit ?:)

Yep, LGTM 👍

**Comment by NN---** [comment]
Thanks.
Then waiting for appveyor.
It takes some time to compile :)

**Comment by NN---** [comment]
Do you know why the build is so slow ?

**Comment by ig-sinicyn** [comment]
Have no idea

**Comment by NN---** [comment]
Why the build failed ?

**Comment by NN---** [comment]
nunit3-console doesn't support net50 yet.

**Review by ig-sinicyn — APPROVED** [review]

**Comment by NN---** [comment]
@ig-sinicyn Thanks for approval.
I see we encountered the new .NET 5.0 breaking change with strings.:(
Need to investigate and fix before merging.

**Comment by NN---** [comment]
Here is what fails.
Need to figure out how to make a better test.
https://github.com/dotnet/dotnet-docker/issues/1877#issuecomment-622125813

**Comment by NN---** [comment]
Found the relevant documentation: https://docs.microsoft.com/en-us/dotnet/standard/base-types/string-comparison-net-5-plus

**Comment by NN---** [comment]
@ig-sinicyn  any idea how to test invariant culture ?
I used ss and ß characters but they are now not the same in 5.0

**Comment by ig-sinicyn** [comment]
> @ig-sinicyn any idea how to test invariant culture ?
> I used ss and ß characters but they are now not the same in 5.0

Feel free to disable tests (with some comment, of course).
 I'm going to scan all codebase for the 
CA1307: Specify StringComparison for clarity
CA1309: Use ordinal StringComparison
after PR completion.

**Comment by NN---** [comment]
I made test which compares ss with ß , validating invariant comparison is used.
But in .NET 5.0 they are not the same even with invariant comparison.
The question do we need to find some other invariant equal character or just remove this case ?:)

**Comment by ig-sinicyn** [comment]
> The question do we need to find some other invariant equal character or just remove this case ?:)

Just remove

**Comment by NN---** [comment]
Found the problem. AppVeyor has .NET 5.0 preview which defined NETCOREAPP3_1 for .NET 5.0 :(
Added a workaround.

**Pull request merged**